horovod/common/operations.cc

```diff
@@ -840,18 +850,231 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
+#if HAVE_CUDA && HOROVOD_GPU_ALLGATHER != 'M'      // 'M' stands for CUDA-aware MPI 
+        bool on_gpu = e.device != CPU_DEVICE_ID;
+        if (on_gpu) {
+          // If shared buffer is not initialized or is not large enough, reallocate
+          if (horovod_global.shared_buffer == nullptr ||
+              horovod_global.shared_buffer_size < total_size * element_size) {
+            if (horovod_global.shared_buffer != nullptr) {
+              MPI_Win_fence(0, horovod_global.window);
+              MPI_Win_free(&horovod_global.window);
+              horovod_global.shared_buffer = nullptr;
+            }
+            int64_t window_size =
+                horovod_global.local_rank == 0 ? total_size * element_size : 0;
+    
+            // Allocate shared memory, give each rank their respective pointer
+            ACTIVITY_START_ALL(entries, timeline, ALLOCATE_SHARED_BUFFER)
+            MPI_Win_allocate_shared(
+                window_size, element_size, MPI_INFO_NULL, horovod_global.local_comm,
+                &horovod_global.shared_buffer, &horovod_global.window);
+    
+            if (horovod_global.local_rank != 0) {
+              int disp_unit;
+              MPI_Aint winsize;
+              MPI_Win_shared_query(horovod_global.window, 0, &winsize, &disp_unit,
+                                   &horovod_global.shared_buffer);
+            }
+            horovod_global.shared_buffer_size = total_size * element_size;
+            ACTIVITY_END_ALL(entries, timeline)
+          }
+
+          cudaStream_t& stream = horovod_global.streams[e.device];
+          CUDA_CHECK(entries, "cudaSetDevice", cudaSetDevice(e.device))
+
+          if (stream == nullptr) {
+            int greatest_priority;
+            CUDA_CHECK(entries, "cudaDeviceGetStreamPriorityRange",
+                       cudaDeviceGetStreamPriorityRange(NULL, &greatest_priority))
+            CUDA_CHECK(entries, "cudaStreamCreateWithPriority",
+                       cudaStreamCreateWithPriority(
+                           &stream, cudaStreamNonBlocking, greatest_priority))
+          }
+          auto event_queue = std::queue<std::pair<std::string, cudaEvent_t>>();
+
+          int64_t local_buffer_len = element_size*recvcounts[horovod_global.rank];
+          void* host_buffer = malloc(local_buffer_len);
+
+          CUDA_CHECK(entries, "cudaMemcpyAsync",
+                   cudaMemcpyAsync(host_buffer, e.tensor->data(),
+                                   local_buffer_len, cudaMemcpyDeviceToHost, stream))
+
+          if (timeline.Initialized()) {
+            RECORD_EVENT(entries, event_queue, MEMCPY_IN_HOST_BUFFER, stream)
+          }
+          WAIT_FOR_EVENTS(entries, timeline, event_queue)
+
+          ACTIVITY_START_ALL(entries, timeline, MPI_ALLGATHER)
+          auto result = MPI_Allgatherv(
+              host_buffer, (int)e.tensor->shape().num_elements(),
+              GetMPIDataType(e.tensor), horovod_global.shared_buffer, recvcounts,
```

Comment:
Yes, and it worked ok, but thinking about it again the behavior could depend on MPI implementation. One option is to only have local rank 0 call the `MPI_Allgatherv`, but at that point it is just like hierarchical allgather except for parallelization. The other option is to `malloc` the allgather output for each local rank, but it feels wasteful.

Reply:
I'm starting to like idea of local rank 0.  We could have every worker copy data into the right spot of the buffer, do a barrier, local_rank=0 does cross-comm allgatherv, everyone does another barrier, profit?

Reply:
Sure! Again, this is almost the same as the hierarchical algorithm except for parallelization. So do you want to make hierarchical allgather the default algorithm and do parallelized version if `is_homogeneous==true`?

Reply:
We should refactor to reuse code between these two as much as we can.  I'm not sure about default yet - our upcoming autotuning should figure it out automatically, so we could keep simple allgatherv default for now.

Reply:
Yes, but if the data is at GPU and `HOROVOD_GPU_ALLGATHER=MPI` is not set then that will fail, right? In the remaining cases we can have simple MPI_Allgatherv as default.

Reply:
Yes, that scenario would have to use the non-hierarchical algorithm with shared buffer and local_rank=0 doing most of the work by default.  The default MPI_Allgatherv is still more economical in terms of overall memory usage on pure CPU since it doesn't require additional shared memory buffer.

Reply:
Ok! I will refactor based on this branching logic then:

Data is at CPU:
If hierarchical_allgather is enabled and is_homogeneous: Copy to shared buffer, MPI_Allgatherv across nodes (with all local ranks), copy to output buffer.
Else: Simple MPI_Allgatherv

Data is at GPU:
If HOROVOD_GPU_ALLGATHER=MPI: Simple MPI_Allgatherv
Else: Copy to CPU shared buffer, MPI_allgatherv across nodes (all local ranks if `is_homogeneous==true`, only local rank 0 otherwise), copy back to GPU

Reply:
I refactored the code, trying to re-use code across different cases as much as possible. There are 4 factors that determine the specific algorithm used: 1. whether data is at CPU or GPU, 2. whether `HOROVOD_GPU_ALLGATHER==MPI`, 3. whether `hierarchical_allgather==true`, 4. whether `is_homogeneous==true`, but the logic is (roughly) as I described in the previous comment. Let me know what you think!